water/lib/Dialect/Wave/IR/WaveInterfaces.cpp

```diff
@@ -476,4 +478,449 @@ llvm::LogicalResult wave::detail::verifyCompatibleOperandsAndResultsOpTrait(
+                                     mlir::Attribute constraintsAttr) {
+  wave::IndexExprsAnalysisInit initObject;
+  initObject.hardwareConstraint =
+      parseWaveConstraints(loc, constraintsAttr, initObject.symbolConstraints);
+  if (initObject.hardwareConstraint == nullptr)
+    return llvm::failure();
+  initObject.wavesPerBlock = initObject.hardwareConstraint.getWavesPerBlock();
+  return initObject;
+}
+
+// Populate `positions` with the positions of thread and GPR-related symbols in
+// the given list.
+static void
+getThreadLikeSymbolPositions(llvm::ArrayRef<mlir::Attribute> symbols,
+                             llvm::SmallVectorImpl<unsigned> &positions) {
+  for (auto &&[i, symbol] : llvm::enumerate(symbols)) {
+    auto indexSymbol = llvm::dyn_cast<wave::WaveIndexSymbolAttr>(symbol);
+    if (!indexSymbol)
+      continue;
+    if (llvm::is_contained({wave::WaveIndexSymbol::THREAD_0,
+                            wave::WaveIndexSymbol::THREAD_1,
+                            wave::WaveIndexSymbol::THREAD_2,
+                            wave::WaveIndexSymbol::GPR_NUMBER},
+                           indexSymbol.getValue()))
+      positions.push_back(i);
+  }
+}
+
+wave::IndexExprsLatticeStorage wave::IndexExprsLatticeStorage::join(
+    const IndexExprsLatticeStorage &lhs, const IndexExprsLatticeStorage &rhs,
+    llvm::ArrayRef<mlir::Attribute> ignoredRhsSymbols) {
+  if (lhs.value == rhs.value)
+    return lhs;
+
+  if (lhs.isTop() || rhs.isTop())
+    return top();
+
+  llvm::StringSet<> ignoredRhsSymbolNames;
+  for (mlir::Attribute attr : ignoredRhsSymbols) {
+    auto symbolAttr = llvm::dyn_cast<wave::WaveSymbolAttr>(attr);
+    if (!symbolAttr)
+      continue;
+    ignoredRhsSymbolNames.insert(symbolAttr.getName());
+  }
+
+  if (lhs.isBottom()) {
+    if (ignoredRhsSymbols.empty() || rhs.isBottom())
+      return rhs;
+
+    llvm::SmallVector<mlir::NamedAttribute> filtered = llvm::filter_to_vector(
+        rhs.getConcreteValue(), [&](mlir::NamedAttribute namedAttr) {
+          return !ignoredRhsSymbolNames.contains(
+              namedAttr.getName().getValue());
+        });
+    return IndexExprsLatticeStorage(mlir::DictionaryAttr::get(
+        rhs.getConcreteValue().getContext(), filtered));
+  }
+
+  if (rhs.isBottom())
+    return lhs;
+
+  mlir::MLIRContext *ctx = lhs.getConcreteValue().getContext();
+  mlir::DictionaryAttr lhsValue = lhs.getConcreteValue();
+  mlir::DictionaryAttr rhsValue = rhs.getConcreteValue();
+
+  llvm::DenseMap<mlir::StringAttr, mlir::Attribute> result;
+  for (mlir::NamedAttribute namedAttr : lhsValue) {
+    result[namedAttr.getName()] = namedAttr.getValue();
+  }
+  for (mlir::NamedAttribute namedAttr : rhsValue) {
+    if (ignoredRhsSymbolNames.contains(namedAttr.getName().getValue())) {
+      continue;
+    }
+
+    auto it = result.find(namedAttr.getName());
+    if (it == result.end()) {
+      result[namedAttr.getName()] = namedAttr.getValue();
+      continue;
+    }
```

Comment:
```suggestion
    auto [it, inserted] = result.try_emplace(namedAttr.getName(), namedAttr.getValue());
    if (inserted)
      continue;
```

Reply:
Change for marginal efficiency gain on copies, but I find `try_emplace` less readable in this context.